### PR TITLE
Decaffeinate

### PR DESCRIPTION
**Pull request by Arcanemagus**
Decaffeinate the code, moving to Standard for linting to follow Atom's [JavaScript Styleguide](https://github.com/atom/atom/blob/master/CONTRIBUTING.md#javascript-styleguide).

In the process I 🔥'd `grunt` as it wasn't really providing any benefit making it worth bringing in as a dependency.

Testing:
* Before 911fb92 was done I ensured that the existing specs passed with the new JS versions of the files, then the specs were converted.
*  The documentation of the `atom/atom` repo at [3482f8f](https://github.com/atom/atom/commit/3482f8fe6f6f5f7145addff7e9b72530e870e3e7) was built using `atomdoc` from this PR with no differences due to this PR. See https://github.com/atom/atomdoc/pull/18#issuecomment-354341911 for details.

_Note: This should be reviewed/merged before #17 as it will greatly simplify that PR._

### CODE REVIEW

File: src/doc.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
+module.exports = class Doc {
```
Comment (50Wliu) [inline src/doc.js]: The class should be on a new line

File: src/utils.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  getLinkMatch (text) {
+    const m = text.match(/\{([\w.]+)\}/)
+    if (m) {
+      return m[1]
+    } else {
+      return null
+    }
+  },
+
+  multiplyString (string, times) {
```
Comment (50Wliu) [inline src/utils.js]: This helper seems...rather unnecessary now that you've converted it to `string.repeat`.

File: src/parser.js [hunk 3]
```diff
@@ -0,0 +1,520 @@
...
+  let visibility = 'Private'
+
+  let rawVisibility = null
+  let rawSummary = tokens[0].text
+  if (rawSummary) {
+    const visibilityMatch = VisibilityRegex.exec(rawSummary)
+    if (visibilityMatch) {
+      visibility = visibilityMatch[1]
+      rawVisibility = visibilityMatch[0]
+      if (rawVisibility) { rawSummary = rawSummary.replace(rawVisibility, '') }
```
Comment (50Wliu) [inline src/parser.js]: If it's an inline if-statement I don't think the braces are necessary.

File: src/parser.js [hunk 4]
```diff
@@ -0,0 +1,520 @@
...
+
+// Public: Parses a docString
+//
+// * `docString` a string from the documented object to be parsed
+//
+// Returns a {Doc} object
+const parse = function (docString) {
+  const lexer = new marked.Lexer()
+  const tokens = lexer.lex(docString)
+  const firstToken = _.first(tokens)
```
Comment (50Wliu) [inline src/parser.js]: `tokens[0]`?

File: src/parser.js [hunk 5]
```diff
@@ -0,0 +1,520 @@
...
+    summary = rawSummary
+    description = generateDescription(tokens, tokenCallback)
+    if (rawVisibility) { description = description.replace(rawVisibility, '') }
+    return {description, summary, visibility}
+  }
+}
+
+var parseArgumentsSection = function (tokens) {
+  const firstToken = _.first(tokens)
+  if (firstToken && (firstToken.type === 'heading')) {
```
Comment (50Wliu) [inline src/parser.js]: No need for extra parentheses

File: src/parser.js [hunk 6]
```diff
@@ -0,0 +1,520 @@
...
+    if (firstToken.text !== 'Arguments' ||
+      firstToken.depth !== SpecialHeadingDepth
```
Comment (50Wliu) [inline src/parser.js]: This can be on the same line as the one before it

File: src/parser.js [hunk 7]
```diff
@@ -0,0 +1,520 @@
...
+      examples.push(example)
+    } else {
+      break
+    }
+  }
+
+  if (examples.length) { return examples }
+}
+
+var parseReturnValues = function (tokens, consumeTokensAfterReturn) {
```
Comment (50Wliu) [inline src/parser.js]: var 👉 const

File: spec/parser-spec.js [hunk 8]
```diff
@@ -0,0 +1,971 @@
+/*
```
Comment (50Wliu) [inline spec/parser-spec.js]: :fire:

File: spec/parser-spec.js [hunk 9]
```diff
@@ -0,0 +1,971 @@
...
+      expect(doc.visibility).toBe('Public')
+      expect(doc.summary).toBe('Batch multiple operations as a single undo/redo step.')
+      expect(doc.description).toBe('Batch multiple operations as a single undo/redo step.')
+      expect(doc.returnValue).not.toBeDefined()
+      expect(doc.examples).not.toBeDefined()
+      expect(doc.delegation).not.toBeDefined()
+    })
+
+    it('parses the description properly', function () {
+      const str = `\
```
Comment (50Wliu) [inline spec/parser-spec.js]: Maybe use `dedent` for all these strings so that the indentation is clearer?

File: src/parser.js (hunk 4 above)
Comment (Arcanemagus) [reply src/parser.js]: That's what I would use everywhere myself, the code was previously using `_.first` everywhere though. I'll happily remove it 😉.

File: src/parser.js (hunk 5 above)
Comment (Arcanemagus) [reply src/parser.js]: Whoops, missed that one. I don't know why `standard` doesn't enable that rule...

File: src/utils.js (hunk 2 above)
Comment (Arcanemagus) [reply src/utils.js]: 🔥'ing it 😉.

**Comment by 50Wliu** [comment]
Conversion generally seems fine, but this still a very large change.  Are there any output discrepancies when generating `atom-api.json` using the Coffeescript version and the JavaScript version?

**Comment by Arcanemagus** [comment]
Okay, finally got around to testing this @50Wliu. The `atom-api.json` generated with the `atomdoc` from this PR _is_ different from the one currently generated by the build process... but that's only because this includes #16 so there are now a bunch of `undefined` in the documentation where previously they were being cut out 😉.

_Note: The "Coffeescript version" and "JavaScript version" stuff is all handled in [donna](https://github.com/atom/donna) and [joanna](https://github.com/atom/joanna). [tello](https://github.com/atom/tello) takes their output, merges it together, and sends it to `atomdoc` to get the final output._

<details>
<summary>atom-api.diff</summary>

```diff
diff --git "a/.\\docs\\output\\atom-api.json" "b/.\\docs\\output\\atom-api-original.json"
index f8b63d3d7..8dc9afe78 100644
--- "a/.\\docs\\output\\atom-api.json"
+++ "b/.\\docs\\output\\atom-api-original.json"
@@ -9587,7 +9587,7 @@
           "returnValues": [
             {
               "type": "Package",
-              "description": "Returns a {Package} or undefined."
+              "description": "Returns a {Package} or ."
             }
           ]
         },
@@ -9653,7 +9653,7 @@
           "returnValues": [
             {
               "type": "Package",
-              "description": "Returns a {Package} or undefined."
+              "description": "Returns a {Package} or ."
             }
           ]
         },
@@ -18782,7 +18782,7 @@
           "returnValues": [
             {
               "type": "DisplayMarkerLayer",
-              "description": "Returns a {DisplayMarkerLayer} or `undefined` if no layer exists with the\ngiven id."
+              "description": "Returns a {DisplayMarkerLayer} or `` if no layer exists with the\ngiven id."
             }
           ]
         },
@@ -18876,7 +18876,7 @@
           "returnValues": [
             {
               "type": "Cursor",
-              "description": "Returns the first matched {Cursor} or undefined"
+              "description": "Returns the first matched {Cursor} or"
             }
           ]
         },
@@ -19776,7 +19776,7 @@
           "returnValues": [
             {
               "type": "Range",
-              "description": "Returns the selected {Range} or `undefined` if the marker is invalid."
+              "description": "Returns the selected {Range} or `` if the marker is invalid."
             }
           ]
         },
@@ -20132,7 +20132,7 @@
             },
             {
               "type": "Boolean",
-              "description": "Returns a {Boolean} or undefined if no non-comment lines had leading\nwhitespace."
+              "description": "Returns a {Boolean} or  if no non-comment lines had leading\nwhitespace."
             }
           ]
         },
@@ -22128,7 +22128,7 @@
           "returnValues": [
             {
               "type": "TextEditor",
-              "description": "Returns a {TextEditor} or `undefined` if the workspace center's current\nactive item is not a {TextEditor}."
+              "description": "Returns a {TextEditor} or `` if the workspace center's current\nactive item is not a {TextEditor}."
             }
           ]
         },
@@ -22208,7 +22208,7 @@
           "returnValues": [
             {
               "type": "Dock",
-              "description": "Returns a {Dock}, the {WorkspaceCenter}, or `undefined` if no item exists\nwith the given URI."
+              "description": "Returns a {Dock}, the {WorkspaceCenter}, or `` if no item exists\nwith the given URI."
             }
           ]
         },
@@ -22230,7 +22230,7 @@
           "returnValues": [
             {
               "type": "Dock",
-              "description": "Returns a {Dock}, the {WorkspaceCenter}, or `undefined` if no item exists\nwith the given URI."
+              "description": "Returns a {Dock}, the {WorkspaceCenter}, or `` if no item exists\nwith the given URI."
             }
           ]
         },
@@ -22252,7 +22252,7 @@
           "returnValues": [
             {
               "type": "Pane",
-              "description": "Returns a {Pane} or `undefined` if no item exists with the given URI."
+              "description": "Returns a {Pane} or `` if no item exists with the given URI."
             }
           ]
         },
@@ -22274,7 +22274,7 @@
           "returnValues": [
             {
               "type": "Pane",
-              "description": "Returns a {Pane} or `undefined` if no pane exists for the given item."
+              "description": "Returns a {Pane} or `` if no pane exists for the given item."
             }
           ]
         },
@@ -23434,7 +23434,7 @@
           "returnValues": [
             {
               "type": "TextEditor",
-              "description": "Returns an {TextEditor} or `undefined` if the current active item is not an\n{TextEditor}."
+              "description": "Returns an {TextEditor} or `` if the current active item is not an\n{TextEditor}."
             }
           ]
         },
```

</details>

**Comment by Arcanemagus** [comment]
Bump 😛.

**Comment by 50Wliu** [comment]
Will try to take another look soon!

**Pull request merged**